golem-worker-executor/src/durable_host/mod.rs

```diff
@@ -2434,6 +2435,243 @@ impl PrivateDurableWorkerState {
+    ) -> Result<(OplogIndex, Tx), Err>
+    where
+        Err: From<GolemError>,
+    {
+        if matches!(
+            *function_type,
+            DurableFunctionType::WriteRemoteTransaction(None)
+        ) {
+            if self.is_live() {
+                let (tx_id, tx) = handler.create_new().await?;
+                self.oplog
+                    .add_and_commit_safe(OplogEntry::begin_remote_transaction(tx_id))
+                    .await
+                    .map_err(GolemError::runtime)?;
+                let begin_index = self.oplog.current_oplog_index().await;
+                Ok((begin_index, tx))
+            } else {
+                let (begin_index, begin_entry) =
+                    crate::get_oplog_entry!(self.replay_state, OplogEntry::BeginRemoteTransaction)?;
+
+                let pre_entry = self
+                    .replay_state
+                    .lookup_oplog_entry_with_condition(
+                        begin_index,
+                        OplogEntry::is_pre_remote_transaction,
+                        OplogEntry::no_concurrent_side_effect,
+                    )
+                    .await;
+
+                let jumped = self
+                    .replay_state
+                    .is_in_skipped_region(begin_index.next())
+                    .await;
+
+                let begin_entry = if jumped {
+                    let (_, entry) = crate::get_oplog_entry!(
+                        self.replay_state,
+                        OplogEntry::BeginRemoteTransaction
+                    )?;
+                    entry
+                } else {
+                    begin_entry
+                };
+
+                let tx_id = try_match!(
+                    begin_entry,
+                    OplogEntry::BeginRemoteTransaction {
+                        timestamp: _,
+                        transaction_id,
+                    }
+                )
+                .map_err(|_| GolemError::runtime("Unexpected oplog entry"))?;
+
+                let (tx_id, tx) = handler.create_replay(&tx_id).await?;
+
+                let mut restart = false;
+
+                if let Some((_pre_index, pre_entry)) = pre_entry {
+                    let end_entry = self
+                        .replay_state
+                        .lookup_oplog_entry_with_condition(
+                            begin_index,
+                            OplogEntry::is_end_remote_transaction,
+                            OplogEntry::no_concurrent_side_effect,
+                        )
+                        .await;
+
+                    // println!(
+                    //     "begin_transaction_function tx_id {tx_id}, pre_index {pre_index} end_entry {} jumped {jumped}",
+                    //     end_entry.clone().map(|(i, _)| i.to_string()).unwrap_or("None".to_string())
+                    // );
+                    if end_entry.is_none()
+                        && pre_entry.is_pre_commit_remote_transaction(begin_index)
```

Comment:
I think I would personally restructure this code in a small way as follows

You are already doing a pre-entry check  before which internally checks  if its a is_pre_commit_remote_transaction. So instead of checking pre_commit and pre_rollback together, I would just just check each of them and handle the logic as separate functions. And obviously avoids this second check, but most improtantly more readable.